doc/ampi/03-using.rst

```diff
@@ -294,6 +294,63 @@ heap allocation because in AMPI, the stack sizes are fixed at the
 beginning (and can be specified from the command line) and stacks do not
 grow dynamically.
 
+Automatic Position-Independent Executable Runtime Relocation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Position-Independent Executable (PIE) Globals allows fully automatic
+privatization of global variables on GNU/Linux systems without
+modification of user code. All languages (C, C++, Fortran, etc.) are
+supported. Runtime migration, load balancing, checkpointing, and SMP
+mode are all fully supported.
+
+This method works by combining a specific method of building binaries
+with GNU extensions to the dynamic linker. First, AMPI's toolchain
+wrapper compiles your user program as a Position-Independent Executable
+(PIE) and links it against a special shim of function pointers instead
+of the normal AMPI runtime. It then builds a small loader utility that
+links directly against AMPI. This loader dynamically opens the PIE
+binary after the AMPI runtime is fully initialized. The glibc
+extension ``dl_iterate_phdr`` is called before and after the ``dlopen``
```

Comment:
Apple is very good about exposing and documenting the structure of binaries, so I believe there is opportunity for the same idea to work on macOS. It would need some exploration. The tlsglobals code I wrote would be a good starting point.